### DISCUSSION TITLE
Got one!, Wifi issue ?

### DISCUSSION THREAD

**Discussion by emilic14** [pcmihnea/HA_WMR500#6] [Q&A] [answered]
Hi there.
So I went on and got one ^^

All modified (but the fw part, as I am not running the same FW '1476' and well I havent gotten to check more that part).

I am seeing the Wifi issue, or more like lack of answering of the MQTT service.
I was wondering if you have looked more on this ?

I do see that a powercycle fixes the issue, which i guess I could solve by having a smart plug toggle the connexion once I detect that the system is no longer answering... 
But do you know if there are other ways to connect to the device and make it reset the server ? as in could I telnet to it? or mqtt reboot it ? 

Just wondering if you have thought of some solutions already.

**Comment by pcmihnea** [comment, marked as answer]
Hi,

I'm amazed it's still available in stores :) Where did you get it from, and how much did it cost?

I'm 99% sure my firmware was the last one available, as I regularly checked for updates while the servers were still up. The test-points are easily accessible, so you can backup and flash my modified firmware dump without much hassle. What FW version does your device report?

The WiFi issue was present for me even before the servers went down - the solution was the same, ie. to power cycle - I don't use batteries, as it's easier to hardware reset it just by unplugging it, plus the onboard statistics aren't of much use to me.

I haven't found much info in terms of open ports/socket, albeit I didn't invest much time on this topic. You could use the reset testpoint present under a sticker underneath the device, next to the battery door (pull it low via a switch, or another MCU).
In case you got the time and skills, dump the firmware or use mine and reverse-engineer it using Ghidra (free) or IdaPro (not free). I can give you hints on the first one.

**Reply by emilic14** [reply]
Got it 2nd hand ^^ actually new, unboxed never used... and for about 30euros plus shipping ^^

It stated that its running FW # 1476

For the time being I have a home assistant connected switch that I can toggle on/off when I detect that its no longer responding, I was just wondering if there was a better solution ^^

I am actually quite curious of how you manage to do all this reverse engineering ! (i am impressed!) and I am more than willing to give it a try to use Ghidra and learn a bit of how it is if you don't mind giving some tips. Don't have a ton of free time nor many skills but hey I managed to follow your instructions and put the device up and running ^^

**Reply by pcmihnea** [reply]
I'll put together some instructions for a Ghidra workspace, and update the readme - stay tuned :)

**Reply by pcmihnea** [reply]
I've updated (and spell checked :) ) the [readme](https://github.com/pcmihnea/HA_WMR500/blob/main/README.md) with a starting point for further reverse-engineering. Hope it helps!
By the way, are there any more available to buy, or was it a one-time sale only?

**Reply by emilic14** [reply]
Thanks for the guide update!!! I'll totally give it a try and let you know.

It was a second hand sell (even if the product was as new), so I would say a one-time only.
I've seen several second hand being sold as the unit is impossible to be configured... I just simply explained the situation to the seller who understood (i hope confirmed what I was saying) and then agreed to sell it for a lower price that the one he was asking initially.

**Reply by emilic14** [reply]
Hello again and once more a huge thanks!!

I've finally had some time to look into this again, and wow I am enjoying it so much!! and learning!

Took me a bit to get the binary... I thought I did not had an adapter to get the jtag info... and then I realized I had a bus pirate. Just to learn that its not really working with SWD devices although it seems it is supported by openocd (only for jtag).
And then I learned that I can use a raspberrypi as an SWD reader with openocd !!!

So I now got access, got the binaries, but still learning the ghidra part ^^
Also to get my binary I used >flash read_bank 0 bank0.bin 
and then >flash read_bank 1 bank1.bin
as the listing reported 2 memory banks...

Well still learning and trying to find the addresses you stated as I am not seeing the same ones ^^
for example : i get address 8004ca57 | Starting WICED v3.5.2 | "\nStarting WICED v3.5.2" | ds

FYI, in section 4 of your readme you state that there are 6 screws holding the cover, there are actually 7 (1 is hidden in the upper middle part of the screen).

**Reply by pcmihnea** [reply]
Hi,

Good to know you're making progress, and ramping up knowledge!

I too get into such situations that no mather how many tools I've racked up so far, there's still a 5$ device I'm missing that's 2~30 days delivery away, if it's even stocked anymore (thanks, chip shortage!).
There might be some differences in terms of code locations if your device has a different version for the firmware (["data"]["6"]["indoor"]["w8"]["c82"]) and/or hardware (["data"]["6"]["indoor"]["w8"]["c83"]) - what are their values? Mine were '1490' and '1' respectively.
Since my patched FW is already in the repo, you could do a visual comparison to see if their structures are similar - maybe something simple as a fixed offset could be an explanation.

Good hint on the hidden screw, I didn't even used it the moment I've opened the case :)
Edit: updated the readme.md to mark the screws more clearly.

**Reply by emilic14** [reply]
well c82 for me is 1476, and c83 is 1.
I cant really compare the files as it seems that my file is not getting disassembled as well as your hex file by ghidra... Not sure what I am doing wrong ^^ but I'll keep on reading on how to use ghidra.
Good news is I can see what you are talking about on your readme when I disassemble your hex file. Although to be honest I don't understand what it is that you are changing (BL to NOP) and how did you get to that decision.

**Reply by pcmihnea** [reply]
So your firmware may be older. Unfortunately, there's no more official ways to upgrade it.
The hex file was exported by ghidra after patching, and by design includes memory mapping (addresses). The raw binary, as the name implies, does not. So to compare you need to convert one to the other, easier hex -> bin.
If your workspace isn't the same, be sure you've mapped the bin to the correct starting address, as per readme.
As for the patching itself, when browsing the disassembled source code in ghidra on the right view, clicking on a code line highlights the machine instructions in the middle view. Then it's just a matter of replacing function calls ( branch to, or bl), and/or call arguments with a different value or just bypassing it with nop ( no operation).
